packages/cdk/lib/riff-raff-yaml.ts

```diff
@@ -0,0 +1,236 @@
+import { writeFileSync } from 'fs';
+import path from 'path';
+import { GuStack } from '@guardian/cdk/lib/constructs/core';
+import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
+import type { App } from 'aws-cdk-lib';
+import { Token } from 'aws-cdk-lib';
+import type { CfnFunction } from 'aws-cdk-lib/aws-lambda';
+import { dump } from 'js-yaml';
+
+interface RiffRaffYaml {
+	regions: Set<string>;
+	allowedStages: Set<string>;
+	deployments: Map<RiffRaffDeploymentName, RiffRaffDeploymentProps>;
+}
+
+type RiffRaffDeploymentName = string;
+
+interface RiffRaffDeploymentProps {
+	type: string;
+	stacks: Set<string>;
+	app: string;
+	contentDirectory: string;
+	parameters: Record<string, string | boolean>;
+	dependencies?: RiffRaffDeploymentName[];
+	actions?: string[];
+}
+
+interface RiffRaffDeployment {
+	name: RiffRaffDeploymentName;
+	props: RiffRaffDeploymentProps;
+}
+
+// TODO `GuLambdaFunction` already takes `fileName` as a prop, can it expose it?
+const getLambdaFileName = (lambda: GuLambdaFunction): string => {
+	const {
+		app,
+		node: { defaultChild },
+	} = lambda;
+
+	const { code } = defaultChild as CfnFunction;
+	const inferredFilename = `${app}.zip`;
+
+	if (Token.isUnresolved(code)) {
+		console.log(
+			`Lambda's code is unresolved. Inferring lambda filename to ${inferredFilename}`,
+		);
+		return inferredFilename;
+	}
+
+	const { s3Key } = code as CfnFunction.CodeProperty;
+	if (!s3Key) {
+		throw new Error('Only S3 based lambdas are supported');
+	}
+	return s3Key.split('/').at(-1) ?? inferredFilename;
+};
+
+export class RiffRaffYamlFile {
+	private readonly riffRaffYaml: RiffRaffYaml;
+	private readonly outDir: string;
+	private readonly guStacks: GuStack[];
+
+	private getRegions(): Set<string> {
+		return new Set(
+			this.guStacks.map(({ region }) => {
+				if (Token.isUnresolved(region)) {
+					throw new Error('Region is a token');
+				}
+				return region;
+			}),
+		);
+	}
+
+	private getStages(): Set<string> {
+		return new Set(this.guStacks.map((_) => _.stage));
+	}
+
+	private getLambdas(stack: GuStack): GuLambdaFunction[] {
+		return stack.node
+			.findAll()
+			.filter(
+				(construct) => construct instanceof GuLambdaFunction,
```

Comment:
If we looked for [`CfnFunction`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda.CfnFunction.html), this could be super generic...